File: kotlin/src/main/kotlin/boj/7000/7511.kt
package boj.`7000`

/*
https://www.acmicpc.net/problem/7511
7511.소셜 네트워킹 어플리케이션
풀이1.1652ms
*/

var disjointSet: IntArray = IntArray(0)
fun find(vertex: Int): Int {
    if (disjointSet[vertex] != vertex)
        disjointSet[vertex] = find(disjointSet[vertex])

    return disjointSet[vertex]
}

fun union(vertex1: Int, vertex2: Int) {
    val root1 = find(vertex1)
    val root2 = find(vertex2)

    if (root1 <= root2)
        disjointSet[root2] = root1
    else
        disjointSet[root1] = root2
}

fun main() {
    val answer = StringBuilder()

    repeat(readln().toInt()) { i ->
        val n = readln().toInt()
        disjointSet = IntArray(n + 1) { it }

        answer.append("Scenario ${i + 1}:\n")

        repeat(readln().toInt()) {
            val (a, b) = readln().split(" ").map { it.toInt() }
            union(a, b)
        }

        repeat(readln().toInt()) {
            val (a, b) = readln().split(" ").map { it.toInt() }
            if (find(a) == find(b))
                answer.append("1\n")
            else
                answer.append("0\n")
        }
        answer.append("\n")
    }

    println(answer)
}

File: python/BOJ/7XXX/7511.py
"""
https://www.acmicpc.net/problem/7511
7511.소셜 네트워킹 어플리케이션
풀이1.952ms
"""
import sys

input = sys.stdin.readline


def find(vertex):
    if disjoint_set[vertex] != vertex:
        disjoint_set[vertex] = find(disjoint_set[vertex])

    return disjoint_set[vertex]


def union(vertex1, vertex2):
    root1 = find(vertex1)
    root2 = find(vertex2)

    if root1 <= root2:
        disjoint_set[root2] = root1
    else:
        disjoint_set[root1] = root2


for i in range(int(input())):
    print("Scenario " + str(i + 1) + ":")
    n = int(input())
    disjoint_set = [i for i in range(n + 1)]

    for _ in range(int(input())):
        a, b = map(int, input().split())
        union(a, b)

    for _ in range(int(input())):
        a, b = map(int, input().split())
        if find(a) == find(b):
            print("1")
        else:
            print("0")
    print()